common/cache/domainCache.go

```diff
@@ -199,39 +191,43 @@ func (c *domainCache) GetAllDomain() map[string]*DomainCacheEntry {
 // WARN: the beforeCallback function will be triggered by domain cache when holding the domain cache lock,
 // make sure the callback function will not call domain cache again in case of dead lock
 // afterCallback will be invoked when NOT holding the domain cache lock.
-func (c *domainCache) RegisterDomainChangeCallback(shard int, initialNotificationVersion int64, beforeCallback CallbackFn, afterCallback CallbackFn) {
+func (c *domainCache) RegisterDomainChangeCallback(shard int, initialNotificationVersion int64,
+	prepareCallback PrepareCallbackFn, callback CallbackFn) {
 	c.Lock()
-	c.beforeCallbacks[shard] = beforeCallback
-	c.afterCallbacks[shard] = afterCallback
-	domainNotificationVersion := c.domainNotificationVersion
+	c.prepareCallbacks[shard] = prepareCallback
+	c.callbacks[shard] = callback
 	c.Unlock()
 
 	// this section is trying to make the shard catch up with domain changes
-	if domainNotificationVersion > initialNotificationVersion {
-		domains := DomainCacheEntries{}
-		for _, domain := range c.GetAllDomain() {
-			domains = append(domains, domain)
-		}
-		// we mush notify the change in a ordered fashion
-		// since history shard have to update the shard info
-		// with domain change version.
-		sort.Sort(domains)
-		for _, domain := range domains {
-			if domain.notificationVersion >= initialNotificationVersion {
-				beforeCallback(nil, domain)
-				afterCallback(nil, domain)
-			}
+	domains := DomainCacheEntries{}
+	for _, domain := range c.GetAllDomain() {
+		domains = append(domains, domain)
+	}
+	// we mush notify the change in a ordered fashion
+	// since history shard have to update the shard info
+	// with domain change version.
+	sort.Sort(domains)
+
+	prevEntries := []*DomainCacheEntry{}
+	nextEntries := []*DomainCacheEntry{}
+	for _, domain := range domains {
+		if domain.notificationVersion >= initialNotificationVersion {
+			prevEntries = append(prevEntries, nil)
```

Comment:
the idea is to match the size of prev entries and next entries.
for catch up case (shard got loaded and try to catch up with domain change), since we do not store a prev version of the domain, we put nil here.